scripts/check_docs_formatted.py

```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/env python3
+"""Check code snippets in docs are formatted by black."""
+import argparse
+import os
+import re
+import textwrap
+from collections.abc import Sequence
+from pathlib import Path
+from re import Match
+
+import black
+from black.mode import Mode, TargetVersion
+from black.parsing import InvalidInput
+
+TARGET_VERSIONS = ["py37", "py38", "py39", "py310", "py311"]
+SNIPPED_RE = re.compile(
+    r"(?P<before>^(?P<indent> *)```\s*python\n)"
+    r"(?P<code>.*?)"
+    r"(?P<after>^(?P=indent)```\s*$)",
+    re.DOTALL | re.MULTILINE,
+)
+
+
+class CodeBlockError(Exception):
+    """A code block parse error."""
+
+
+def format_str(
+    src: str,
+    black_mode: black.FileMode,
+) -> tuple[str, Sequence[CodeBlockError]]:
+    """Format a single docs file string."""
+    errors: list[CodeBlockError] = []
+
+    def _snipped_match(match: Match[str]) -> str:
+        code = textwrap.dedent(match["code"])
+        try:
+            code = black.format_str(code, mode=black_mode)
+        except InvalidInput as e:
+            errors.append(CodeBlockError(e))
+
+        code = textwrap.indent(code, match["indent"])
+        return f'{match["before"]}{code}{match["after"]}'
+
+    src = SNIPPED_RE.sub(_snipped_match, src)
+    return src, errors
+
+
+def format_file(
+    file: Path,
+    black_mode: black.FileMode,
+    error_known: bool,  # noqa: FBT001
+    args: argparse.Namespace,
+) -> int:
+    """Check the formatting of a single docs file."""
+    with file.open() as f:
+        contents = f.read()
+
+    # Remove everything before the first example
+    contents = contents[contents.find("## Example") :]
+
+    # Remove everything after the last example
+    contents = contents[: contents.rfind("```")] + "```"
+
+    new_contents, errors = format_str(contents, black_mode)
+
+    if errors and not args.skip_errors and not error_known:
+        for error in errors:
+            rule_name = file.name.split(".")[0]
+            print(f"Docs parse error for {rule_name} docs: {error}")
+
+        return 2
+
+    if contents != new_contents:
+        rule_name = file.name.split(".")[0]
+        print(
+            f"Rule {rule_name} docs are not formatted. This section should be"
+            " rewritten to:",
+        )
+
+        # Add indentation so that snipped can be copied directly to docs
+        for line in new_contents.splitlines():
+            output_line = "///"
+            if len(line) > 0:
+                output_line = f"{output_line} {line}"
+
+            print(output_line)
+
+        print("\n")
+
+        return 1
+
+    return 0
+
+
+def main(argv: Sequence[str] | None = None) -> int:
+    """Check code snippets in docs are formatted by black."""
+    parser = argparse.ArgumentParser(
+        description="Check code snippets in docs are formatted by black.",
+    )
+    parser.add_argument("--skip-errors", action="store_true")
+    parser.add_argument("--generate-docs", action="store_true")
```

Comment:
Is my understanding correct that this defaults to `false` (we don't generate the docs in the CI)?

Reply:
I like it!